File: solutions/haskell/sum-of-multiples/1/src/SumOfMultiples.hs
module SumOfMultiples (sumOfMultiples) where

import Data.List (nub)

sumOfMultiples :: [Integer] -> Integer -> Integer
sumOfMultiples factors limit = sum . nub . concat $ [build c limit | c <- factors]
  where
    build :: Integer -> Integer -> [Integer]
    build c limit  
             | c /=0 = filter (\x -> x `mod` c == 0) [1 .. limit - 1]
             | True = [0]

File: solutions/elixir/sum-of-multiples/1/lib/sum_of_multiples.ex
defmodule SumOfMultiples do
  @doc """
  Adds up all numbers from 1 to a given end number that are multiples of the factors provided.
  """
  @spec to(non_neg_integer, [non_neg_integer]) :: non_neg_integer

  def to(limit, factors) do
    do_collect(limit, factors, [])
    |> Enum.reduce(MapSet.new(), fn x, acc -> MapSet.union(acc, MapSet.new(x)) end)
    |> Enum.sum()
  end

  def do_collect(_limit, [], acc), do: acc

  def do_collect(limit, [h | t], acc) do
    do_collect(limit, t, [do_to(limit, h, [h]) | acc])
  end

  defp do_to(limit, base, [h | _] = acc) do
    cond do
      base > limit -> []
      base == 0 -> [base]
      limit > h + base -> do_to(limit, base, [h + base | acc])
      true -> acc
    end
  end
end
